### PR TITLE
JTJ 47 goals

### PR DESCRIPTION
**Pull request by tgi-wtag**
In This PR i worked on [this Story](https://welltravel.atlassian.net/browse/JTJ-47?atlOrigin=eyJpIjoiZDEwOTI4YTEwNDc4NDQ2YWEyOWUwNGJhNTJmMTBmZjMiLCJwIjoiaiJ9)
where i added the functionality to create show and update Goals.

https://github.com/tgi-wtag/Journey-tg/assets/141820846/0bd64cef-3f20-48e1-a681-2e62a666395a

### CODE REVIEW

File: app/views/goals/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<h1><%= t('goal.edit') %></h1>
+
+ <%= form_for @goal do |f| %>
```
Comment (meb-wtag) [inline app/views/goals/edit.html.erb:3]: please use simple_form

File: app/views/goals/index.html.erb [hunk 2]
```diff
@@ -1 +1,25 @@
...
+    <% end %>
+<% current_user.goals.each do |goal| %>
+  <div class="content-card">
+    <%= link_to goal_path(goal), class: "goal-link" do %>
+      <h2><%= goal.description %></h2>
+      <p><%= goal.status %></p>
+    <% end %>
+  </div>
+<% end %>
+
```
Comment (meb-wtag) [inline app/views/goals/index.html.erb:22]: unnecessary line

File: app/views/goals/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+</div>
+
+<div class="field">
+  <%= f.label :deadline %>
+  <%= f.date_field :deadline %>
+</div>
+
+<div class="field">
+  <%= f.label :status %>
+  <%= f.select :status, options_for_select([['Setted', 'setted'], ['Working On', 'working_on'], ['Archived', 'archived']]) %>
```
Comment (meb-wtag) [inline app/views/goals/edit.html.erb:16]: please always the same

File: spec/controllers/goals_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
+require 'rails_helper'
+
+RSpec.describe GoalsController, type: :controller do
+  let(:user) { create(:user) }
+
+  before do
+    user = create(:user)
```
Comment (meb-wtag) [inline spec/controllers/goals_controller_spec.rb:7]: remove line

File: app/controllers/goals_controller.rb [hunk 5]
```diff
@@ -2,4 +2,50 @@ class GoalsController < ApplicationController
   def goals
     @goals = Goal.all
   end
+
+  def new
+    @goal = Goal.new
+  end
+
+  def create
+    @goal = Goal.new(goal_params)
```
Comment (chm-wtag) [inline app/controllers/goals_controller.rb:11]: this is not needed

File: app/controllers/goals_controller.rb [hunk 6]
```diff
@@ -2,4 +2,50 @@ class GoalsController < ApplicationController
...
+    @goal = Goal.find(params[:id])
+    @goal.deadline
+  end
+
+  def edit
+    @goal = Goal.find(params[:id])
+  end
+
+  def update
+    @goal = Goal.find(params[:id])
```
Comment (chm-wtag) [inline app/controllers/goals_controller.rb:37]: use before_action and DRY this code

**Review by chm-wtag — APPROVED** [review]

File: app/views/goals/edit.html.erb (hunk 3 above)
Comment (tgi-wtag) [reply app/views/goals/edit.html.erb]: thank you for your comment

File: spec/factories/goal.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :goal do
+    description { FFaker::Lorem.sentence }
+    deadline { FFaker::Time.date }
+    status { 'setted' }
```
Comment (evk-wtag) [inline spec/factories/goal.rb:5]:
```suggestion
    status { :setted }
```
or you can use `sample` to randomly get a key

File: spec/support/authentication_helper.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+module AuthenticationHelper
+  def sign_in_user(user)
+    session[:user_id] = user.id
+  end
+end
```
Comment (evk-wtag) [inline spec/support/authentication_helper.rb:5]: Check this PR: https://github.com/mla-wtag/ml_Journey/pull/6/files for an easier implementation of the helper method.

**Review by evk-wtag — APPROVED** [review]

File: spec/controllers/goals_controller_spec.rb (hunk 4 above)
Comment (tgi-wtag) [reply spec/controllers/goals_controller_spec.rb:7]: i did thank you

File: spec/controllers/goals_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    it 'fails to create a goal with missing description' do
+      post :create, params: { goal: { description: '', status: 'setted' } }
+      expect(response).to render_template(:new)
+    end
+  end
+
+  describe 'GET #show' do
+    let(:goal) { create(:goal, user: user) }
+
+    it 'returns a successful response' do
```
[the remark this answers is not in the document]
Comment (tgi-wtag) [reply spec/controllers/goals_controller_spec.rb:42]: then there is no successful response ;)

File: spec/controllers/goals_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      get :show, params: { id: goal.id }
+      expect(response).to have_http_status(:ok)
+    end
+  end
+
+  describe 'GET #edit' do
+    let(:goal) { create(:goal, user: user) }
+
+    it 'returns a successful response' do
```
[the remark this answers is not in the document]
Comment (tgi-wtag) [reply spec/controllers/goals_controller_spec.rb:51]: it doesnt return a successful response :)

File: spec/controllers/goals_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+  before do
+    user = create(:user)
+    sign_in_user(user)
+  end
+
+  describe 'GET #index' do
+    it 'returns a successful response' do
+      get :index
+      expect(response).to have_http_status(:ok)
```
[the remark this answers is not in the document]
Comment (tgi-wtag) [reply spec/controllers/goals_controller_spec.rb:14]: i think this works to right ? if index isnt found it wont have http status ok? or am i wrong here?

File: spec/models/goal_spec.rb [hunk 12]
```diff
@@ -2,8 +2,12 @@
...
 RSpec.describe Goal, type: :model do
   it 'is valid with valid attributes' do
-    goal = Goal.new(description: 'This is an example goal')
+    goal = build(:goal)
     expect(goal).to be_valid
   end
-  it { should validate_presence_of(:description) }
+
+  it 'is not valid without a description' do
+    goal = build(:goal, description: nil)
```
[the remark this answers is not in the document]
Comment (tgi-wtag) [reply spec/models/goal_spec.rb:10]: thank you i did, insightfull saved 3 lines of code, thank you

File: app/controllers/goals_controller.rb (hunk 5 above)
Comment (tgi-wtag) [reply app/controllers/goals_controller.rb:11]: thank you i deleted it

File: spec/support/authentication_helper.rb (hunk 8 above)
Comment (tgi-wtag) [reply spec/support/authentication_helper.rb:5]: thank you i did

File: app/controllers/goals_controller.rb [hunk 13]
```diff
@@ -1,5 +1,52 @@
...
+        flash[:error] = t('errors.no_description')
+        redirect_to new_goal_path
+      end
+    else
+      flash[:error] = t('errors.creating_failed')
+      render :new
+    end
+  end
+
+  def show
```
Comment (lra-wtag) [inline app/controllers/goals_controller.rb:29]:
```suggestion
  def show; end
```

File: app/controllers/goals_controller.rb [hunk 14]
```diff
@@ -1,5 +1,52 @@
...
+  end
+
+  def edit
```
Comment (lra-wtag) [inline app/controllers/goals_controller.rb:32]:
```suggestion
  def edit: end
```

File: db/migrate/20231013083210_alter_table_goals.rb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+class AlterTableGoals < ActiveRecord::Migration[7.0]
+  def up
+    add_reference :goals, :user, foreign_key: true
+  end
+
+  def down
+    remove_reference :goals, :user, foreign_key: true
```
Comment (lra-wtag) [inline db/migrate/20231013083210_alter_table_goals.rb:7]:
```suggestion
    remove_reference :goals, :user, foreign_key: false
```
right ?

**Review by lra-wtag — CHANGES_REQUESTED** [review]

File: db/migrate/20231013083210_alter_table_goals.rb (hunk 15 above)
Comment (tgi-wtag) [reply db/migrate/20231013083210_alter_table_goals.rb:7]: yess

File: app/models/goal.rb [hunk 16]
```diff
@@ -1,4 +1,5 @@
 class Goal < ApplicationRecord
+  belongs_to :user
```
Comment (lra-wtag) [inline app/models/goal.rb:2]: missing spec for this association

**Review by lra-wtag — CHANGES_REQUESTED** [review]

File: app/models/goal.rb (hunk 16 above)
Comment (tgi-wtag) [reply app/models/goal.rb:2]: oh thank you for the review, i added it.

**Review by lra-wtag — APPROVED** [review]

**Pull request merged**